File: GLLara/GLLResourceManager.m
//
//  GLLResourceManager.m
//  GLLara
//
//  Created by Torsten Kammer on 04.09.12.
//  Copyright (c) 2012 Torsten Kammer. All rights reserved.
//

#import "GLLResourceManager.h"

#import "GLLModel.h"
#import "GLLPreferenceKeys.h"
#import "GLLTexture.h"

#import "GLLara-Swift.h"

struct GLLAlphaTestBlock
{
    GLuint mode;
    GLfloat reference;
};

@interface GLLResourceManager ()
{
    NSMutableDictionary *shaders;
    NSMutableDictionary<GLLShaderData*, GLLModelProgram *> *programs;
    NSMutableDictionary *textures;
    NSMutableDictionary *models;
    NSMutableDictionary<NSString *, GLLPipelineStateInformation*> *pipelines;
    NSMutableDictionary<NSString *, id<MTLFunction>> *functions;
}

- (NSData *)_dataForFilename:(NSString *)filename error:(NSError *__autoreleasing*)error;
- (NSString *)_utf8StringForFilename:(NSString *)filename error:(NSError *__autoreleasing*)error;
- (id)_valueForKey:(id)key from:(NSMutableDictionary *)dictionary ifNotFound:(id(^)(void))supplier;
- (id<MTLFunction>)_functionForName:(NSString *)name shader:(GLLShaderData*)shader error:(NSError *__autoreleasing*)error;
- (void)_recreateSampler;

@end

static GLLResourceManager *sharedManager;

@implementation GLLResourceManager

+ (GLLResourceManager *)sharedResourceManager
{
    if (!sharedManager)
        sharedManager = [[GLLResourceManager alloc] init];
    
    return sharedManager;
}

- (id)init
{
    if (!(self = [super init])) return nil;
    
    _metalDevice = MTLCreateSystemDefaultDevice();
    
    shaders = [[NSMutableDictionary alloc] init];
    programs = [[NSMutableDictionary alloc] init];
    textures = [[NSMutableDictionary alloc] init];
    models = [[NSMutableDictionary alloc] init];
    
    _library = [_metalDevice newDefaultLibrary];
    _pixelFormat = MTLPixelFormatBGRA8Unorm;
    _depthPixelFormat = MTLPixelFormatDepth32Float;
    
    [self _recreateSampler];
    
    MTLDepthStencilDescriptor *depthDescriptor = [[MTLDepthStencilDescriptor alloc] init];
    depthDescriptor.depthCompareFunction = MTLCompareFunctionLessEqual;
    depthDescriptor.depthWriteEnabled = YES;
    _normalDepthStencilState = [_metalDevice newDepthStencilStateWithDescriptor:depthDescriptor];
    
    // Alpha test buffers
    struct GLLAlphaTestBlock alphaBlock = { .mode = 1, .reference = .9 };
    _alphaTestPassGreaterBuffer = [_metalDevice newBufferWithBytes:&alphaBlock length:sizeof(alphaBlock) options:MTLResourceStorageModeManaged];
    _alphaTestPassGreaterBuffer.label = @"alpha-test-pass-greater";
    
    struct GLLAlphaTestBlock alphaBlockPassLess = { .mode = 2, .reference = .9 };
    _alphaTestPassLessBuffer = [_metalDevice newBufferWithBytes:&alphaBlockPassLess length:sizeof(alphaBlock) options:MTLResourceStorageModeManaged];
    _alphaTestPassLessBuffer.label = @"alpha-test-pass-less";
    
    [[NSUserDefaultsController sharedUserDefaultsController] addObserver:self forKeyPath:[@"values." stringByAppendingString:GLLPrefAnisotropyAmount] options:NSKeyValueObservingOptionNew context:0];
    [[NSUserDefaultsController sharedUserDefaultsController] addObserver:self forKeyPath:[@"values." stringByAppendingString:GLLPrefUseAnisotropy] options:NSKeyValueObservingOptionNew context:0];
    
    return self;
}

- (void)dealloc;
{
    [models.allValues makeObjectsPerformSelector:@selector(unload)];
    [textures.allValues makeObjectsPerformSelector:@selector(unload)];
    [programs.allValues makeObjectsPerformSelector:@selector(unload)];
    [shaders.allValues makeObjectsPerformSelector:@selector(unload)];
    
    models = nil;
    textures = nil;
    programs = nil;
    shaders = nil;
}

#pragma mark - Retrieving resources

- (GLLModelDrawData *)drawDataForModel:(GLLModel *)model error:(NSError *__autoreleasing*)error;
{
    return [self _valueForKey:model.baseURL from:models ifNotFound:^{
        return [[GLLModelDrawData alloc] initWithModel:model resourceManager:self];
    }];
}

/*- (GLLModelProgram *)programForDescriptor:(GLLShaderData *)description error:(NSError *__autoreleasing*)error;
{
    NSParameterAssert(description);
    
    return [self _valueForKey:description from:programs ifNotFound:^{
        return [[GLLModelProgram alloc] initWithDescriptor:description resourceManager:self error:error];
    }];
}*/

- (GLLTexture *)textureForURL:(NSURL *)textureURL error:(NSError *__autoreleasing*)error;
{
    return [self _valueForKey:textureURL from:textures ifNotFound:^{
        NSURL *effectiveURL = textureURL;
        NSData *textureData = [NSData dataWithContentsOfURL:textureURL options:NSDataReadingUncached error:error];
        if (!textureData)
        {
            // Second attempt: Maybe there is a default version of that in the bundle.
            // If not, then keep error from first read.
            effectiveURL = [[NSBundle mainBundle] URLForResource:textureURL.lastPathComponent withExtension:nil];
            if (!effectiveURL)
                return (GLLTexture *) nil;
        }
        return [[GLLTexture alloc] initWithURL:effectiveURL device:self.metalDevice error:error];
    }];
}

/*- (GLLShader *)shaderForName:(NSString *)shaderName additionalDefines:(NSDictionary *)defines usedTexCoords:(NSIndexSet *)texCoords type:(GLenum)type error:(NSError *__autoreleasing*)error;
{
    NSParameterAssert(shaderName);
    NSParameterAssert(defines);
    
    NSDictionary *key = @{ @"name" : shaderName,
                           @"defines": defines,
                           @"texCoords": texCoords
    };
    
    return [self _valueForKey:key from:shaders ifNotFound:^{
        NSString *shaderSource = [self _utf8StringForFilename:shaderName error:error];
        if (!shaderSource) return (GLLShader *) nil;
        
        // Actual loading
        return [[GLLShader alloc] initWithSource:shaderSource name:shaderName additionalDefines:defines usedTexCoords:texCoords type:type error:error];
    }];
}*/

- (GLLPipelineStateInformation *)pipelineForVertex:(GLLVertexAttribAccessorSet *)vertexDescriptor shader:(GLLShaderData *)shader error:(NSError *__autoreleasing*)error; {
    NSParameterAssert(vertexDescriptor);
    NSParameterAssert(shader);
    
    // TODO Does this work?
    NSDictionary *key = @{
        @"shader": shader,
        @"vertexDescriptor": vertexDescriptor.vertexDescriptor
    };
    
    return [self _valueForKey:key from:pipelines ifNotFound:(id)^{
        id<MTLFunction> vertexFunction = [self _functionForName:shader.vertexName shader:shader error:error];
        if (!vertexFunction) {
            return (id)nil;
        }
        
        id<MTLFunction> fragmentFunction = [self _functionForName:shader.fragmentName shader:shader error:error];
        if (!fragmentFunction) {
            return (id)nil;
        }
        
        MTLRenderPipelineDescriptor *descriptor = [[MTLRenderPipelineDescriptor alloc] init];
        descriptor.vertexFunction = vertexFunction;
        descriptor.fragmentFunction = fragmentFunction;
        descriptor.colorAttachments[0].pixelFormat = self.pixelFormat;
        descriptor.depthAttachmentPixelFormat = self.depthPixelFormat;
        descriptor.vertexDescriptor = vertexDescriptor.vertexDescriptor;
        
        id<MTLRenderPipelineState> renderPipelineState = [self.metalDevice newRenderPipelineStateWithDescriptor:descriptor error:error];
        if (!renderPipelineState) {
            return (id)nil;
        }
        
        GLLPipelineStateInformation *information = [[GLLPipelineStateInformation alloc] init];
        information.vertexProgram = vertexFunction;
        information.fragmentProgram = fragmentFunction;
        information.pipelineState = renderPipelineState;
        return (id)information;
    }];
}

- (id<MTLBuffer>)squareVertexArray
{
    if (!_squareVertexArray)
    {
        float coords[] = {
            -1.0f, -1.0f,
            1.0f, -1.0f,
            -1.0f, 1.0f,
            1.0f, 1.0f
        };
        _squareVertexArray = [_metalDevice newBufferWithBytes:coords length:sizeof(coords) options:MTLResourceStorageModePrivate];
        _squareVertexArray.label = @"square-vertex";
    }
    return _squareVertexArray;
}

#pragma mark - OpenGL limits

- (NSInteger)maxAnisotropyLevel
{
    return 16; // TODO
}

#pragma mark - Testing

- (void)clearInternalCaches;
{
    [models.allValues makeObjectsPerformSelector:@selector(unload)];
    [textures.allValues makeObjectsPerformSelector:@selector(unload)];
    [programs.allValues makeObjectsPerformSelector:@selector(unload)];
    [shaders.allValues makeObjectsPerformSelector:@selector(unload)];
    
    [models removeAllObjects];
    [textures removeAllObjects];
    [programs removeAllObjects];
    [shaders removeAllObjects];
}

-(void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSKeyValueChangeKey,id> *)change context:(void *)context {
    if (object == [NSUserDefaultsController sharedUserDefaultsController]) {
        [self _recreateSampler];
    } else {
        [super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
    }
}

#pragma mark - Private methods

- (id)_valueForKey:(id)key from:(NSMutableDictionary *)dictionary ifNotFound:(id(^)(void))supplier;
{
    NSParameterAssert(key);
    id result = dictionary[key];
    if (!result)
    {
        result = supplier();
        dictionary[key] = result;
    }
    return result;
    
}

- (NSData *)_dataForFilename:(NSString *)filename error:(NSError *__autoreleasing*)error;
{
    NSString *actualFilename = [[filename componentsSeparatedByString:@"\\"] lastObject];
    
    NSURL *localURL = [NSURL URLWithString:actualFilename relativeToURL:nil];
    NSData *localData = [NSData dataWithContentsOfURL:localURL];
    if (localData) return localData;
    
    NSURL *resourceURL = [NSURL URLWithString:actualFilename relativeToURL:[[NSBundle mainBundle] resourceURL]];
    return [NSData dataWithContentsOfURL:resourceURL options:0 error:error];
}
- (NSString *)_utf8StringForFilename:(NSString *)filename error:(NSError *__autoreleasing*)error;
{
    NSData *data = [self _dataForFilename:filename error:error];
    if (!data) return nil;
    
    return [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
}

- (id<MTLFunction>)_functionForName:(NSString *)name shader:(GLLShaderData*)shader error:(NSError *__autoreleasing*)error {
    NSParameterAssert(name);
    NSParameterAssert(shader);
    
    NSDictionary *key = @{
        @"name": name,
        @"shader": shader
    };
    
    return [self _valueForKey:key from:pipelines ifNotFound:(id)^{
        MTLFunctionConstantValues *values = [[MTLFunctionConstantValues alloc] init];
        bool *valuesArray = calloc(sizeof(bool), GLLFunctionConstantBoolMax);
        NSIndexSet *setParameters = shader.activeBoolConstants;
        for (NSUInteger i = 0; i < GLLFunctionConstantBoolMax; i++) {
            if ([setParameters containsIndex:i]) {
                valuesArray[i] = true;
            }
        }
        [values setConstantValues:valuesArray type:MTLDataTypeBool withRange:NSMakeRange(0, GLLFunctionConstantBoolMax)];
        //free(valuesArray);
        
        // TODO Make this dependent on what is actually going on in the scene
        int oneLight = 1;
        [values setConstantValue:&oneLight type:MTLDataTypeInt atIndex:GLLFunctionConstantNumberOfUsedLights];
        
        return [self.library newFunctionWithName:name constantValues:values error:error];
    }];
}

- (void)_recreateSampler {
    MTLSamplerDescriptor *samplerDescriptor = [[MTLSamplerDescriptor alloc] init];
    samplerDescriptor.minFilter = MTLSamplerMinMagFilterLinear;
    samplerDescriptor.magFilter = MTLSamplerMinMagFilterLinear;
    samplerDescriptor.mipFilter = MTLSamplerMipFilterLinear;
    samplerDescriptor.rAddressMode = MTLSamplerAddressModeRepeat;
    samplerDescriptor.sAddressMode = MTLSamplerAddressModeRepeat;
    samplerDescriptor.tAddressMode = MTLSamplerAddressModeRepeat;
    samplerDescriptor.supportArgumentBuffers = YES;
    
    BOOL useAnisotropy = [[NSUserDefaults standardUserDefaults] boolForKey:GLLPrefUseAnisotropy];
    NSInteger anisotropyAmount = [[NSUserDefaults standardUserDefaults] integerForKey:GLLPrefAnisotropyAmount];
    samplerDescriptor.maxAnisotropy = useAnisotropy ? anisotropyAmount : 0;
    
    _metalSampler = [_metalDevice newSamplerStateWithDescriptor:samplerDescriptor];
}

@end

File: GLLara/GLLResourceManager.swift
//
//  GLLResourceManager.swift
//  GLLara
//
//  Created by Torsten Kammer on 07.07.22.
//  Copyright © 2022 Torsten Kammer. All rights reserved.
//

import Foundation
import Metal
import AppKit
import Combine

/*
 * Stores all resources for the program.
 */
@objc class GLLResourceManager: NSObject {
    
    @objc static var shared: GLLResourceManager = GLLResourceManager()
    
    override init() {
        metalDevice = MTLCreateSystemDefaultDevice()!
        library = metalDevice.makeDefaultLibrary()!
        pixelFormat = .bgra8Unorm
        depthPixelFormat = .depth32Float
        
        let normalDepthStencilDescriptor = MTLDepthStencilDescriptor()
        normalDepthStencilDescriptor.depthCompareFunction = .less
        normalDepthStencilDescriptor.isDepthWriteEnabled = true
        normalDepthStencilDescriptor.label = "normal depth mode"
        normalDepthStencilState = metalDevice.makeDepthStencilState(descriptor: normalDepthStencilDescriptor)!
        
        let copyDepthDescriptor = MTLDepthStencilDescriptor()
        copyDepthDescriptor.depthCompareFunction = .greaterEqual
        copyDepthDescriptor.isDepthWriteEnabled = true;
        copyDepthDescriptor.label = "depth copy mode"
        depthStencilStateForCopy = metalDevice.makeDepthStencilState(descriptor: copyDepthDescriptor)!
    
        // Can be rendered as triangle stripe
        let squareCoords: [Float32] = [
            -1.0, -1.0,
            1.0, -1.0,
            -1.0, 1.0,
            1.0, 1.0
        ]
        squareVertexArray = metalDevice.makeBuffer(bytes: squareCoords, length: MemoryLayout<Float32>.stride * squareCoords.count, options: .storageModeManaged)!
        squareVertexArray.label = "square-vertex"
        
        let squarePipelineDescriptor = MTLRenderPipelineDescriptor()
        squarePipelineDescriptor.vertexFunction = library.makeFunction(name: "squareVertex")!
        squarePipelineDescriptor.fragmentFunction = library.makeFunction(name: "squareFragment")!
        squarePipelineDescriptor.colorAttachments[0].pixelFormat = pixelFormat
        squarePipelineDescriptor.colorAttachments[0].isBlendingEnabled = true
        squarePipelineDescriptor.colorAttachments[0].alphaBlendOperation = .max
        squarePipelineDescriptor.colorAttachments[0].sourceRGBBlendFactor = .sourceAlpha
        squarePipelineDescriptor.colorAttachments[0].sourceAlphaBlendFactor = .one
        squarePipelineDescriptor.colorAttachments[0].destinationRGBBlendFactor = .oneMinusSourceAlpha
        squarePipelineDescriptor.colorAttachments[0].destinationAlphaBlendFactor = .one
        squarePipelineDescriptor.label = "square"
        squarePipelineState = try! metalDevice.makeRenderPipelineState(descriptor: squarePipelineDescriptor)
        
        let copyDepthPipelineDescriptor = MTLRenderPipelineDescriptor()
        copyDepthPipelineDescriptor.vertexFunction = library.makeFunction(name: "copyDepthVertex")!
        copyDepthPipelineDescriptor.fragmentFunction = library.makeFunction(name: "copyDepthFragment")!
        copyDepthPipelineDescriptor.colorAttachments[0].pixelFormat = pixelFormat
        copyDepthPipelineDescriptor.depthAttachmentPixelFormat = depthPixelFormat
        copyDepthPipelineDescriptor.label = "copyDepth"
        copyDepthPipelineState = try! metalDevice.makeRenderPipelineState(descriptor: copyDepthPipelineDescriptor)
        
        let checkDepthPipelineDescriptor = MTLRenderPipelineDescriptor()
        checkDepthPipelineDescriptor.vertexFunction = library.makeFunction(name: "depthBufferCheckVertex")!
        checkDepthPipelineDescriptor.fragmentFunction = library.makeFunction(name: "depthBufferCheckFragment")!
        checkDepthPipelineDescriptor.colorAttachments[0].pixelFormat = pixelFormat
        checkDepthPipelineDescriptor.label = "checkDepth"
        depthBufferCheckState = try! metalDevice.makeRenderPipelineState(descriptor: checkDepthPipelineDescriptor)
        
        let drawHudPipelineDescriptor = MTLRenderPipelineDescriptor()
        drawHudPipelineDescriptor.vertexFunction = library.makeFunction(name: "hudTextDrawerVertex")!
        drawHudPipelineDescriptor.fragmentFunction = library.makeFunction(name: "hudTextDrawerFragment")!
        drawHudPipelineDescriptor.colorAttachments[0].pixelFormat = pixelFormat
        drawHudPipelineDescriptor.colorAttachments[0].isBlendingEnabled = true
        // Texture for this is premultiplied
        drawHudPipelineDescriptor.colorAttachments[0].sourceRGBBlendFactor = .sourceAlpha
        drawHudPipelineDescriptor.colorAttachments[0].sourceAlphaBlendFactor = .sourceAlpha
        drawHudPipelineDescriptor.colorAttachments[0].destinationRGBBlendFactor = .oneMinusSourceAlpha
        drawHudPipelineDescriptor.colorAttachments[0].destinationAlphaBlendFactor = .oneMinusSourceAlpha
        drawHudPipelineDescriptor.label = "checkDepth"
        drawHudPipelineState = try! metalDevice.makeRenderPipelineState(descriptor: drawHudPipelineDescriptor)
        
        super.init()
        
        NSUserDefaultsController.shared.addObserver(self, forKeyPath: ("values." + GLLPrefAnisotropyAmount), context: nil)
        NSUserDefaultsController.shared.addObserver(self, forKeyPath: ("values." + GLLPrefUseAnisotropy), context: nil)
        
        recreateSampler()
    }
    
    // Shared programs and buffers that everyone needs sometime
    @objc let maxAnisotropyLevel = 16
    let metalDevice: MTLDevice
    
    let squareVertexArray: MTLBuffer
    let squarePipelineState: MTLRenderPipelineState
    let copyDepthPipelineState: MTLRenderPipelineState
    let depthBufferCheckState: MTLRenderPipelineState // Only used for debugging
    let drawHudPipelineState: MTLRenderPipelineState
    
    let library: MTLLibrary
    let pixelFormat: MTLPixelFormat
    let depthPixelFormat: MTLPixelFormat
    let normalDepthStencilState: MTLDepthStencilState
    let depthStencilStateForCopy: MTLDepthStencilState
    
    // Can and will change if user settings change
    var metalSampler: MTLSamplerState! = nil
    
    func drawDataFuture(model: GLLModel) -> Future<GLLModelDrawData, Error> {
        return futureValue(key: model.baseURL, from: &models, lock: modelsLock) {
            await GLLModelDrawData(model: model, resourceManager: self)
        }
    }
    
    func drawDataAsync(model: GLLModel) async throws -> GLLModelDrawData {
        return try await drawDataFuture(model: model).value
    }
    
    func textureFuture(url: URL) throws -> Future<GLLTexture, Error> {
        futureValue(key: url, from: &textures, lock: texturesLock) {
            var effectiveUrl = url
            do {
                // TODO shouldn't we use this data somewhere? Maybe? Why are we reading this twice?
                _ = try Data(contentsOf: url)
            } catch  {
                // Second attempt: Maybe there is a default version of that in the bundle.
                // If not, then keep error from first read.
                let originalError = error
                let bundleUrl = Bundle.main.url(forResource: url.lastPathComponent, withExtension: nil)
                guard let bundleUrl = bundleUrl else {
                    throw originalError
                }
                effectiveUrl = bundleUrl
            }
            return try GLLTexture(url: effectiveUrl, device: self.metalDevice)
        }
    }
    
    func textureAsync(url: URL) async throws -> GLLTexture {
        return try await textureFuture(url: url).value
    }
    
    @objc func texture(url: URL) throws -> GLLTexture {
        return try throwingRunAndBlockReturn {
            try await self.textureAsync(url: url)
        }
    }
    
    func pipeline(vertex: GLLVertexAttribAccessorSet, shader: GLLShaderData, numberOfTexCoordSets: Int, texCoordAssignments: [Int:Int], hasVariableBoneWeights: Bool = false) throws -> GLLPipelineStateInformation {
        // TODO Does this work?
        let key: [String : AnyHashable] = [
            "shader": shader,
            "vertexDescriptor": vertex.vertexDescriptor,
            "numTexCoords": numberOfTexCoordSets,
            "assignments": texCoordAssignments,
            "variableBoneWeights": hasVariableBoneWeights
        ]
        
        return try pipelinesLock.withLock {
            return try value(key: key, from: &pipelines) {
                let vertexFunction = try function(name: shader.vertexName!, shader: shader, numberOfTexCoordSets: numberOfTexCoordSets, texCoordAssignments: texCoordAssignments, hasVariableBoneWeights: hasVariableBoneWeights)
                let fragmentFunction = try function(name: shader.fragmentName!, shader: shader, numberOfTexCoordSets: numberOfTexCoordSets, texCoordAssignments: texCoordAssignments, hasVariableBoneWeights: hasVariableBoneWeights)
                
                let descriptor = MTLRenderPipelineDescriptor()
                
                descriptor.vertexFunction = vertexFunction
                descriptor.fragmentFunction = fragmentFunction
                descriptor.colorAttachments[0].pixelFormat = pixelFormat
                descriptor.colorAttachments[0].isBlendingEnabled = false
                descriptor.depthAttachmentPixelFormat = depthPixelFormat
                descriptor.vertexDescriptor = vertex.vertexDescriptor
                
                let pipelineState = try metalDevice.makeRenderPipelineState(descriptor: descriptor)
                return GLLPipelineStateInformation(pipelineState: pipelineState, vertexProgram: vertexFunction, fragmentProgram: fragmentFunction)
            }
        }
    }

    // Specifically used for testing
    func clearInternalCaches() {
        textures.removeAll()
        models.removeAll()
        pipelines.removeAll()
        functions.removeAll()
    }
    
    private let texturesLock = NSLock()
    private var textures: [URL: Future<GLLTexture, Error>] = [:]
    private let modelsLock = NSLock()
    private var models: [URL: Future<GLLModelDrawData, Error>] = [:]
    private var pipelinesLock = NSLock()
    private var pipelines: [AnyHashable: GLLPipelineStateInformation] = [:]
    private var functions: [AnyHashable: MTLFunction] = [:]
    
    private func function(name: String, shader: GLLShaderData, numberOfTexCoordSets: Int, texCoordAssignments: [Int:Int], hasVariableBoneWeights: Bool) throws -> MTLFunction {
        // TODO does this work?
        let key: [String : AnyHashable] = [
            "name": name,
            "shader": shader,
            "numTexCoords": numberOfTexCoordSets,
            "assignments": texCoordAssignments,
            "variableBoneWeights": hasVariableBoneWeights
        ]
        
        return try value(key: key, from: &functions) {
            let constantValues = MTLFunctionConstantValues()
            
            // Assign bools for active things
            var valuesArray: [Bool] = Array(repeating: false, count: GLLFunctionConstant.boolMax.rawValue)
            let setParameters = shader.activeBoolConstants
            for i in 0..<valuesArray.count {
                if setParameters.contains(i) {
                    valuesArray[i] = true
                }
            }
            constantValues.setConstantValues(valuesArray, type: .bool, range: 0 ..< valuesArray.count)
            
            // Not handled as a feature because it's part of the model file
            var variableBoneWeights = hasVariableBoneWeights
            constantValues.setConstantValue(&variableBoneWeights, type: .bool, index: GLLFunctionConstant.hasVariableBoneWeights.rawValue)
            
            // Assign value for tex coord
            var numTexCoords32 = Int32(numberOfTexCoordSets)
            constantValues.setConstantValue(&numTexCoords32, type: .int, index: GLLFunctionConstant.numberOfTexCoordSets.rawValue)
            
            // Assign values for tex coord assignments
            var indexValuesArray = Array.init(repeating: Int32(0), count: Int(GLLFragmentArgumentIndexTextureMax.rawValue))
            for i in 0..<indexValuesArray.count {
                if let assigned = texCoordAssignments[i] {
                    indexValuesArray[i] = Int32(assigned)
                }
            }
            constantValues.setConstantValues(indexValuesArray, type: .int, range: 100 ..< 100 + indexValuesArray.count)
            
            // Set number of active lights - TODO adjust to what it should really be
            var lights = Int32(3)
            constantValues.setConstantValue(&lights, type: .int, index: GLLFunctionConstant.numberOfUsedLights.rawValue)
            
            var depthPeelActive: Bool = shader.alphaBlending
            constantValues.setConstantValue(&depthPeelActive, type: .bool, index: GLLFunctionConstant.hasDepthPeelFrontBuffer.rawValue)
            
            return try library.makeFunction(name: name, constantValues: constantValues)
        }
    }
    
    private func recreateSampler() {
        let samplerDescriptor = MTLSamplerDescriptor()
        samplerDescriptor.minFilter = .linear
        samplerDescriptor.magFilter = .linear
        samplerDescriptor.mipFilter = .linear
        samplerDescriptor.rAddressMode = .repeat
        samplerDescriptor.sAddressMode = .repeat
        samplerDescriptor.tAddressMode = .repeat
        samplerDescriptor.supportArgumentBuffers = true
        samplerDescriptor.label = "default"
        
        let useAnisotropy = UserDefaults.standard.bool(forKey: GLLPrefUseAnisotropy)
        let anisotropyAmount = UserDefaults.standard.integer(forKey: GLLPrefAnisotropyAmount)
        samplerDescriptor.maxAnisotropy = useAnisotropy ? anisotropyAmount : 0
        
        metalSampler = metalDevice.makeSamplerState(descriptor: samplerDescriptor)
    }
    
    private func value<K, V>(key: K, from dictionary: inout [K: V], ifNotFound: () throws -> V) rethrows -> V {
        if let existing = dictionary[key] {
            return existing
        }
        
        let newItem = try ifNotFound()
        dictionary[key] = newItem
        return newItem
    }
    
    private func futureValue<K, V>(key: K, from dictionary: inout [K: Future<V, Error>], lock: NSLock, create: @Sendable @escaping () async throws -> V) -> Future<V, Error> {
        lock.withLock {
            if let existing = dictionary[key] {
                return existing
            }
            
            let future = Future<V, Error> { promise in
                Task {
                    do {
                        let value = try await create()
                        promise(Result.success(value))
                    } catch {
                        promise(Result.failure(error))
                    }
                }
            }
            dictionary[key] = future
            return future
        }
    }
    
    override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        if object is NSUserDefaultsController {
            recreateSampler()
        } else {
            super.observeValue(forKeyPath: keyPath, of: object, change: change, context: context)
        }
    }
}
